docs/getting-started/part-3/custom-webservice.md

```diff
@@ -0,0 +1,172 @@
+
+## Get Data Frame
+In this section we will learn how we can avoid sending the requests twice to the API using the execution phase information provided by the smart data lake. We will now implement a simple *if ... else* statement that allows us to simply define the schema of the expected response in the **Init** phase and to actually query the data in the **Exec** phase. We start by defining the *schema* instance variable below
+```scala
+private var schema : Option[ArrayType] = None
+```
+in the beginning of the class `CustomWebserviceDataObject`. The previously discussed logic is enclosed in two `// REPLACE BLOCK` comments. Find this section and replace it with the code below
+```scala
+if(context.phase == ExecutionPhase.Init){
+  import org.apache.spark.sql.types._
+  // define schema
+  schema = Some(ArrayType(StructType(Array(
+    StructField("icao24", StringType),
+    StructField("firstSeen", IntegerType),
+    StructField("estDepartureAirport", StringType),
+    StructField("lastSeen", IntegerType),
+    StructField("estArrivalAirport", StringType),
+    StructField("callsign", StringType),
+    StructField("estDepartureAirportHorizDistance", IntegerType),
+    StructField("estDepartureAirportVertDistance", IntegerType),
+    StructField("estArrivalAirportHorizDistance", IntegerType),
+    StructField("estArrivalAirportVertDistance", IntegerType),
+    StructField("departureAirportCandidatesCount", IntegerType),
+    StructField("arrivalAirportCandidatesCount", IntegerType)
+  ))))
+
+  // simply return an empty data frame
+  session.emptyDataFrame
+} else {
+  // place the new implementation of currentQueryParameters below     
+
+  // given the query parameters, generate all requests
+  val departureRequests = currentQueryParameters.map(
+    param => s"${baseUrl}?airport=${param.airport}&begin=${param.begin}&end=${param.end}"
+  )
+  // make request
+  val departuresResponses = departureRequests.map(request(_))
+
+  // create dataframe with the correct schema and add created_at column with the current timestamp
+  val departuresDf = departuresResponses.toDF("responseBinary")
+    .withColumn("responseString", byte2String($"responseBinary"))
+    .select(from_json($"responseString", schema.get).as("response"))
+    .withColumn("created_at", current_timestamp())
+    .select(explode($"response").as("record"), $"created_at")
+    .select("record.*", "created_at")
+  
+  // put simple nextState logic below
+
+  departuresDf
+}
+```
+We see that we we need to so some transformations to flatten the result returned by the API. Spark has lots of *User-Defined Functions* (short **udf**) that can be used out of the box. We used such a column based function *from_json* to parse the response string with the right schema.
+
+:::tip
+The return type of the response is `Array[Byte]`. To convert that to `Array[String]` the *udf* function `byte2String `has been used. This function is a nice example how one can define such *udfs* by oneselfs.
```

Comment:
I guess you should define that URL in the tutorial then? It s not included in the file CustomWebserviceDataObject-1.scala so as a reader I'm confused here and I was actually googling byte2String to see if it was a predefined udf :D